### PR TITLE
fixed dependencies and overhaul to csv filter level

### LINKED ISSUE #1: plugin doesn't work with logstash >= 5
the plugin still locks the dependency on logstash-core-plugin-api at ~> 1, it needs a [change like the one performed in the supported plugins](https://github.com/logstash-plugins/logstash-filter-date/commit/0e9649b8b643bf03f83d8106021c59e5f4279ae0)

### PR DESCRIPTION
**Pull request by colinsurprenant**
Fixes #1 

- Fixed `logstash-core-plugin-api` dependency to make it work on recent logstash versions
- General overhaul to the level of the csv filter
- Fixed specs using old API
- Added encoding specs

### Goal
The csv codec is a nice addition when paired with the file input/output for instance when stateful information such as columns are relevant per-file which is not possible to manage using the csv filter. 

### TODO
- [x] docs
- [ ] add support for `:target` option (new PR followup)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,5 @@
+## 1.0.0
+  - Fixed dependencies and overhaul to csv filter level, fixed & more specs.
```
Comment (yaauie) [inline CHANGELOG.md]: the phrase _"to csv filter level"_ isn't particularly clear to me.

```suggestion
  - Fixed dependencies to work with Logstash X.Y
  - Overhauled to match features of the CSV Filter
  - Improved spec coverage
```

File: lib/logstash/codecs/csv.rb [hunk 2]
```diff
@@ -51,113 +56,119 @@ class LogStash::Codecs::CSV < LogStash::Codecs::Base
...
 
     # validate conversion types to be the valid ones.
-    @convert.each_pair do |column, type|
-      if !VALID_CONVERT_TYPES.include?(type)
-        raise LogStash::ConfigurationError, "#{type} is not a valid conversion type."
-      end
+    bad_types = @convert.values.select do |type|
+      !CONVERTERS.has_key?(type.to_sym)
+    end.uniq
+    raise(LogStash::ConfigurationError, "Invalid conversion types: #{bad_types.join(', ')}") unless bad_types.empty?
```
Comment (yaauie) [inline lib/logstash/codecs/csv.rb:98]: +1 to error message covering _all_ bad types together

File: lib/logstash/codecs/csv.rb [hunk 3]
```diff
@@ -51,113 +56,119 @@ class LogStash::Codecs::CSV < LogStash::Codecs::Base
...
+
+    # @convert_symbols contains the symbolized types to avoid symbol conversion in the transform method
+    @convert_symbols = @convert.inject({}){|result, (k, v)| result[k] = v.to_sym; result}
```
Comment (yaauie) [inline lib/logstash/codecs/csv.rb]: (totally a nitpick thread)

Are we constrained to Ruby 2.1+ (e.g., via running only on Logstash 6+?)

If so, this reads a bit clearer:

```suggestion
    @convert_symbols = @convert.map { | k, v| [k, v.to_sym] }.to_h
```

Comment (yaauie) [reply lib/logstash/codecs/csv.rb]: Another possibly more clear option is `Enumerable#each_with_object`, which takes away the need for the accumulator to be returned from each iteration of the block.

```diff
-    @convert_symbols = @convert.inject({}){|result, (k, v)| result[k] = v.to_sym; result}
+    @convert_symbols = @convert.each_with_object({}){|(k, v), result| result[k] = v.to_sym}
```

File: lib/logstash/codecs/csv.rb [hunk 4]
```diff
@@ -51,113 +56,119 @@ class LogStash::Codecs::CSV < LogStash::Codecs::Base
...
-            next if ignore_field?(index)
-            field_name =  ( !@columns[index].nil? ? @columns[index] : "column#{(index+1)}")
-            value      = fields
-          elsif fields.is_a?(Array) # Got headers
-            field_name = fields[0]
-            value      = fields[1]
+      values = CSV.parse_line(data, :col_sep => @separator, :quote_char => @quote_char)
+
+      if (@autodetect_column_names && @columns.empty?)
+        @columns = values
```
Comment (yaauie) [inline lib/logstash/codecs/csv.rb:117]: nitpick: would it be worth logging that we have detected the columns?

**Review by yaauie — COMMENTED** [review]
:+1::tada: I've left a couple nitpicky comments, but this is great. So many of the CSV filter options really only make sense in the context of a single stream of data, and make so much more sense in the codec scope.

File: lib/logstash/codecs/csv.rb (hunk 4 above)
Comment (colinsurprenant) [reply lib/logstash/codecs/csv.rb:117]: we don't in the csv filter but I guess we could ... at debug level?

**Comment by colinsurprenant** [comment]
- pushed fixes per @yaauie's review
- pushed docs updates /cc @karenzone

**Comment by karenzone** [comment]
FYI, this plugin isn't included in the Logstash Reference yet.  Is that as expected? 
It **is** included in the Versioned Plugin Reference. 

In case you're wondering, I tested the docs by dropping the content into the skeleton of a different codec.

**Review by karenzone — APPROVED** [review]
Builds cleanly and LGTM.

**Comment by colinsurprenant** [comment]
@karenzone thanks a lot. 
I guess that's a question of promoting that plugin from a prototype state to a proper v1.0 state. Once this is merge we can look into that.

**Comment by colinsurprenant** [comment]
@yaauie Anything else review-wise?

**Review by yaauie — APPROVED** [review]
LGTM 🎉

**Comment by colinsurprenant** [comment]
v1.0.0 published!

**Pull request merged**